### PR TITLE
Sms otp pw recovery

### PR DESCRIPTION
**Pull request by RushanNanayakkara**
### Proposed changes in this pull request

- Add new governance configs for notification based password recovery.
- Handle partial updates of the newly introduced configs. 
- Add multi attribute support for password recovery via RecoveryAPI V2.

### When should this PR be merged

Dependent PR : 
- https://github.com/wso2/carbon-identity-framework/pull/5744

### Related Issues
- https://github.com/wso2-enterprise/iam-product-management/issues/51
- https://github.com/wso2-enterprise/iam-engineering/issues/534

## Related PRs 
- https://github.com/wso2/identity-api-server/pull/621
- https://github.com/wso2/carbon-identity-framework/pull/5744
- https://github.com/wso2/product-is/pull/20557

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java [hunk 1]
```diff
@@ -1,17 +1,19 @@
 /*
- * Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ * Copyright (c) 2024, WSO2 LLC. (http://www.wso2.com).
```
Comment (Avarjana) [inline components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java:2]:
```suggestion
 * Copyright (c) 2016-2024, WSO2 LLC. (http://www.wso2.com).
```

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java [hunk 2]
```diff
@@ -88,10 +99,12 @@ public void updateConfiguration(String tenantDomain, Map<String, String> configu
             residentIdp.setFederatedAuthenticatorConfigs(configsToSave.toArray(new
                     FederatedAuthenticatorConfig[configsToSave.size()]));
             identityProviderManager.updateResidentIdP(residentIdp, tenantDomain);
+        } catch (IdentityProviderManagementClientException e) {
+            log.debug("Client error while updating identityManagement Properties of Resident Idp.", e);
```
Comment (Avarjana) [inline components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java:103]:
```suggestion
            log.debug("Client error while updating identityManagement properties of Resident IdP.", e);
```

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java [hunk 3]
```diff
@@ -88,10 +99,12 @@ public void updateConfiguration(String tenantDomain, Map<String, String> configu
...
+            throw new IdentityGovernanceClientException(e.getMessage(), e);
```
Comment (Avarjana) [inline components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java:104]: Do we send a proper message in the exception itself?

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java [hunk 4]
```diff
@@ -322,4 +335,62 @@ private boolean isEmailOTPConnector(String connectorName, ConnectorConfig connec
...
+                                isSmsOtpCurrentlyEnabled );
+                configurationDetails.put(EMAIL_LINK_PASSWORD_RECOVERY_PROPERTY,
+                        String.valueOf(enableEmailLinkPasswordRecovery));
+                configurationDetails.put(SMS_OTP_PASSWORD_RECOVERY_PROPERTY,
+                        String.valueOf(enableSmsOtpPasswordRecovery));
+                configurationDetails.put(RECOVERY_NOTIFICATION_PASSWORD_PROPERTY,
+                        String.valueOf(enableEmailLinkPasswordRecovery || enableSmsOtpPasswordRecovery));
+            } else {
+                // This is the scenario where the RECOVERY_NOTIFICATION_PASSWORD_PROPERTY is being disabled.
+                configurationDetails.put(EMAIL_LINK_PASSWORD_RECOVERY_PROPERTY, "false");
```
Comment (Avarjana) [inline components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java:391]: Can we get "false" into a constant. I think there might be existing ones defined.

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java (hunk 3 above)
Comment (RushanNanayakkara) [reply components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java:104]: Yes
example : 
Screenshot 2024-06-19 at 18 08 28

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java (hunk 1 above)
Comment (RushanNanayakkara) [reply components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java]: Addressed with https://github.com/wso2-extensions/identity-governance/pull/824/commits/4bf70b014ad62570f070ab6f94a24ff480a06c6e

**Review by Avarjana — APPROVED** [review]

**Review by sumedhe — APPROVED** [review]

**Pull request merged**